### ISSUE TITLE
Support KEDA scalers

### ISSUE THREAD

**Issue by caniszczyk** [artifacthub/hub#1227]
The KEDA project (keda.sh) wants to make it easier to easier to browse scalers: https://keda.sh/docs/2.2/scalers/

Ideally, I think we can have artifacthub list these and then the KEDA team could either just link to them directly OR embed artifacthub somehow in their website :)?

Reference issues:

https://github.com/kedacore/keda-docs/issues/413
https://github.com/kedacore/keda-docs/issues/412

**Comment by tegioz** [comment]
Hi @caniszczyk @tomkerkhove

This sounds great, KEDA scalers would be a nice addition to Artifact Hub 🙂

Let me explain a bit how Artifact Hub indexes content.

Any organization or user can add repositories to Artifact Hub. At the moment we support several repositories kinds (Helm charts and plugins, OLM, operators, Falco rules, OPA policies, Tinkerbell actions, Krew plugins and Tekton tasks). The `tracker` component polls them periodically and collects metadata as needed. Depending on the repository kind, metadata is extracted one way or another. For kinds that have already defined how a repository should be structured and served, like Helm charts or OLM operators, we have [specialized tracker sources](https://github.com/artifacthub/hub/tree/master/internal/tracker/source). This allowed publishers to start adding existing repositories without requiring any additional work. For other kinds, we have a generic tracker source based on our own [metadata file](https://github.com/artifacthub/hub/blob/master/docs/metadata/artifacthub-pkg.yml).
 
For KEDA scalers, I'd suggest we use the generic tracker source, but we're happy to go the other way if you think it'd be a better fit, it's not a problem at all.

The generic tracker source expects packages' metadata (in this case each KEDA scaler would be a package) to be organized in a specific way (please see [Tinkerbell actions repositories doc](https://artifacthub.io/docs/topics/repositories/#tinkerbell-actions-repositories) as an example, as it'd be quite similar). Each of the KEDA scalers repositories would need to provide some medatata about its packages following these conventions. I guess there would be one repository provided by the **KEDA organization** for the **built-in** scalers, and possibly more from other orgs/users providing external scalers.

We've already prepared Artifact Hub to support KEDA scalers this way. So if you are OK with it, as soon as the metadata for the built-in scalers is ready, we can start listing them 🚀  We've setup a [repository with some sample metadata](https://github.com/tegioz/keda-builtin-scalers/tree/main/artifacthub) for 3 scalers to give it a try. I guess you'd like to automate the metadata generation for all of them somehow though.

Some screenshots of the process of adding the sample repository to Artifact Hub and how the packages would look like:

<img width="622" alt="add-repo" src="https://user-images.githubusercontent.com/1213902/113857263-97127f00-97a2-11eb-8b99-73b628483931.png">

<img width="2017" alt="search" src="https://user-images.githubusercontent.com/1213902/113857274-9bd73300-97a2-11eb-9104-c3188b487b30.png">

<img width="2017" alt="package" src="https://user-images.githubusercontent.com/1213902/113857286-9e398d00-97a2-11eb-8da3-ddd9f11c4587.png">

**Comment by caniszczyk** [comment]
Let's see if this works for @tomkerkhove and team, I think they have some feature requests on filtering and discoverability that they can share

**Comment by tomkerkhove** [comment]
Wow you are fast @tegioz 😄 

Adding @zroubalik who has some concerns.

For external scalers, we are 100% onboard, but we are not sure about the built-in ones if this would be the perfect fit. One option would be to embed it on our website but we are not sure yet if that's ideal for us.

In terms of the filtering, I think the majority of them are already supported but our initial thinking is open on https://github.com/kedacore/keda-docs/issues/412, feel free to let me know what you think and what we can make happen or not. Ideally there would be filters on the left for these.

**Comment by tomkerkhove** [comment]
@caniszczyk for https://github.com/kedacore/keda-docs/issues/413 I think Artifact Hub might be less ideal as it would be polluting it with our integrations?

**Comment by tegioz** [comment]
Just in case it helps, [Tinkerbell](https://docs.tinkerbell.org) is integrating with Artifact Hub this way:

<img width="1601" alt="tinkerbell" src="https://user-images.githubusercontent.com/1213902/113987708-004fcc00-984f-11eb-90ce-a36de592bbd6.png">

Which takes users to the search landing page filtering by Tinkerbell actions:

<img width="2017" alt="tinkerbell2" src="https://user-images.githubusercontent.com/1213902/113988041-53c21a00-984f-11eb-9e31-a4a962d6cc12.png">

One of the main benefits of having the built-in scalers in Artifact Hub would be discoverability. Users searching for `kafka`, for example, may find a Helm chart, an OLM operator or a KEDA scaler, which might drive more users to the project. When viewing a package's detail, a scaler related to the package might show up in the related packages section on the right bottom of the page as well.

But listing only external scalers is fine too, whatever works best for you 🙂 Another option would be to list the built-in ones in in Artifact Hub and keep them in your website as they are as well.

Related to kedacore/keda-docs#413, we can add any kind specific fields you may need, we do that for the other kinds as well. This may include a field to distinguish between built-in or external, or support for trigger authentication. So please don't hesitate requesting any of them.

If you are happy with the metadata structure proposed for KEDA scalers, please let us know and we'll go ahead and merge. At that point Artifact Hub would be able to accept scalers repositories. We've added [some documentation](https://github.com/artifacthub/hub/blob/tegioz/keda/docs/repositories.md#keda-scalers-repositories) that details how to add KEDA repositories to Artifact Hub. This would be eventually displayed on https://artifacthub.io/docs once deployed.

**Comment by zroubalik** [comment]
I think that Artifact Hub is a great place for our Helm Chart and OLM manifest (does it support referencing just installation yaml ?) as well as for external scalers.

my 2c: 

But I don't think that it is good idea for built-in scalers - there aren't any specific packages or artifact related to those scalers, it is realy just documentation issue. I don't see a benefit for users as the documentation and related data will be on two places (AH and our website). And from maintenance perspective this is extra work for us. 

I want to have all KEDA releated documentation on place (keda.sh website) so it is convenient for users and not spread across multiple sources. 

It would be a different story if we are able to integrate AH directly into our website, so we can benefit from its features and have the scalers doc in one place.

**Comment by tomkerkhove** [comment]
I agree on that, if we can embed it, then we can go end-to-end.

**Comment by tomkerkhove** [comment]
For the rest, I'm pretty much liking the full approach though so would love to be able to use it.

Can we merge the current set with focus on just external scalers? We can then already publish our HTTP add-on and Azure Durable Function one as a test. WDYT @zroubalik?

**Comment by zroubalik** [comment]
@tomkerkhove agree! 
And would be nice to add KEDA OLM resources next to the Helm Chart as well
https://github.com/kedacore/keda-olm-operator  
https://operatorhub.io/operator/keda

**Comment by tegioz** [comment]
Awesome, we'll merge this later today as soon as we finish a couple of things. I'll let you know when it's deployed 👍

**Comment by tegioz** [comment]
> And would be nice to add KEDA OLM resources next to the Helm Chart as well
> https://github.com/kedacore/keda-olm-operator
> https://operatorhub.io/operator/keda

Please take a look at the [artifacthub.io/recommendations](https://artifacthub.io/docs/topics/annotations/helm/) annotation. It can help letting users know about other packages once they land in one of them.

<img width="2012" alt="Screen Shot 2021-04-09 at 11 30 39" src="https://user-images.githubusercontent.com/1213902/114160263-087b3a80-9927-11eb-8eeb-afe396075ded.png">

**Comment by tegioz** [comment]
Related to embedding @tomkerkhove @zroubalik, how would you envision this working?

We have another issue open (#1228) related to embedding, but I think the goal of it would be to be able to embed a single Artifact Hub package on a site displaying some details about it nicely, similarly to how Tweets are usually embedded. When we have this feature implemented you could build a page that lists all of them in a single page, taking users to Artifact Hub for further interaction. But I'm not sure if this is what you have in mind for embedding or you are thinking about something else.

Please note that you also have available an [API](https://artifacthub.github.io/hub/api/#/Packages) to interact with the packages available in artifacthub.io, just in case it helps with any kind of integration (built-in or external scalers that are added eventually to Artifact Hub).

**Comment by tomkerkhove** [comment]
For our case we would need to list all built-in on one page similar to https://keda.sh/docs/2.2/scalers/.

We could use the API, but reason why we've reached out to @caniszczyk is because we don't have web dev experience so we'd be stuck there but maybe that's something we can tackle.

**Comment by tegioz** [comment]
Do you mean just the list @tomkerkhove, like in the screenshot below, or also the readme content? If it's just the list, you could leave it as it is initially but linking each of the scalers to its Artifact Hub page. Once we have the embed feature ready, you could make it look nicer by adding some `<blockquote></blockquote>` blocks. We could also generate some embed code for all packages in a repo, to make it easier to embed the full list without having to maintain it manually one by one.

<img width="1020" alt="list" src="https://user-images.githubusercontent.com/1213902/114174118-767c2d80-9938-11eb-82b8-b85910d316c3.png">

**Comment by tomkerkhove** [comment]
We can definately do that, if we know how to do it 😅

But maybe we can maintain this list in parallel to the Artifact Hub specification that we have, but we'd need help for this on our docs though.

**Comment by tegioz** [comment]
We'd be happy to help, that's not a problem 🙂

**Comment by tomkerkhove** [comment]
> > And would be nice to add KEDA OLM resources next to the Helm Chart as well
> > [kedacore/keda-olm-operator](https://github.com/kedacore/keda-olm-operator)
> > [operatorhub.io/operator/keda](https://operatorhub.io/operator/keda)
> 
> Please take a look at the [artifacthub.io/recommendations](https://artifacthub.io/docs/topics/annotations/helm/) annotation. It can help letting users know about other packages once they land in one of them.
> 
> <img alt="Screen Shot 2021-04-09 at 11 30 39" width="2000" src="https://user-images.githubusercontent.com/1213902/114160263-087b3a80-9927-11eb-8eeb-afe396075ded.png">

Can you write me an email on kerkhove.tom@gmail.com about this please? I need to check who owns that organization as it seems I'm no longer part of it.

I will then set up a repo for our metadata for our external scalers

**Comment by tegioz** [comment]
Sorry, that was me testing locally, nothing to worry about. I used it to make the screenshots look nicer 😉 

AKAIK it doesn't exist yet on artifacthub.io, so you should be able to register it.

**Comment by tegioz** [comment]
Once the [images are built](https://github.com/artifacthub/hub/actions/runs/733000003) we'll deploy to staging, run some tests and then we'll deploy to production. I'll ping you when it's done!

**Comment by tomkerkhove** [comment]
> Sorry, that was me testing locally, nothing to worry about. I used it to make the screenshots look nicer 😉
> 
> AKAIK it doesn't exist yet on artifacthub.io, so you should be able to register it.

I'm just asking since https://artifacthub.io/packages/search?page=1&repo=kedacore is listed and I don't know if this was because I PRed it in or somebody created an org for it.

**Comment by tomkerkhove** [comment]
It's through the old Helm registry I think, ok.

#TGIF

**Comment by tegioz** [comment]
Oh, I see. That's from the initial import we did from the Helm Hub. But you can [claim its ownership](https://artifacthub.io/docs/topics/repositories/#ownership-claim), many orgs are doing it.

**Comment by tegioz** [comment]
Deployed! 🚀 You can now add KEDA scalers repositories 😄 Please let us know if you run into any issue. Documentation is available [here](https://artifacthub.io/docs/topics/repositories/#keda-scalers-repositories).

We've added [this repository](https://github.com/tegioz/keda-builtin-scalers) to the `staging` environment to give it a try. You can [check it out](https://staging.artifacthub.io/packages/search?page=1&repo=keda-scalers-test) to see how yours will look like. If you search (in `staging`) for something like `kafka`, the kafka scaler should show up in the top results among some available charts.

**Comment by tomkerkhove** [comment]
Thanks a lot! I'll craft a list of our current external scalers and add it to our organization onces it's in production!

**Comment by tegioz** [comment]
Awesome! It's just been deployed to production.
